clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -428,11 +428,19 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
         Attrs.insert(A);
       else if (auto *A = FD->getAttr<ReqdWorkGroupSizeAttr>())
         Attrs.insert(A);
+      // Allow the following intel kernel attributes only on the lambda
+      // (function object) function, that is called directly from a kernel
+      // (i.e. the one passed to the parallel_for function). Emit a warning and
+      // ignore all other cases.
+      else if (auto *A = FD->getAttr<SYCLIntelNumSimdWorkItemsAttr>()) {
+        if (ParentFD == SYCLKernel) {
+          Attrs.insert(A);
+        } else {
+          SemaRef.Diag(A->getLocation(), diag::warn_attribute_ignored) << A;
+          FD->dropAttr<SYCLIntelKernelArgsRestrictAttr>();
+        }
+      }
       else if (auto *A = FD->getAttr<SYCLIntelKernelArgsRestrictAttr>()) {
```

Comment:
Since more lambda-applicable attributes are going to be added, would it make sense to somehow factor this out into a separate template function?